statistics/handle/handle_hist.go

```diff
@@ -147,70 +148,78 @@ func (h *Handle) SubLoadWorker(ctx sessionctx.Context, exit chan struct{}, exitW
 			}
 		}
 	}()
+	var lastTask *NeededColumnTask
 	for {
-		err := h.handleOneTask(readerCtx, ctx.(sqlexec.RestrictedSQLExecutor), exit)
+		task, err := h.HandleOneTask(lastTask, readerCtx, ctx.(sqlexec.RestrictedSQLExecutor), exit)
+		lastTask = task
 		if err != nil {
 			switch err {
 			case errExit:
 				return
 			default:
-				time.Sleep(10 * time.Millisecond)
+				time.Sleep(h.Lease() / 10)
 				continue
 			}
 		}
 	}
 }
 
-// handleOneTask handles one column task.
-func (h *Handle) handleOneTask(readerCtx *statsReaderContext, ctx sqlexec.RestrictedSQLExecutor, exit chan struct{}) (err error) {
+// HandleOneTask handles one column task.
+func (h *Handle) HandleOneTask(lastTask *NeededColumnTask, readerCtx *StatsReaderContext, ctx sqlexec.RestrictedSQLExecutor, exit chan struct{}) (task *NeededColumnTask, err error) {
 	defer func() {
 		// recover for each task, worker keeps working
 		if r := recover(); r != nil {
 			buf := make([]byte, 4096)
 			stackSize := runtime.Stack(buf, false)
 			buf = buf[:stackSize]
-			logutil.BgLogger().Error("stats loading panicked", zap.String("stack", string(buf)))
+			logutil.BgLogger().Error("stats loading panicked", zap.Any("error", r), zap.String("stack", string(buf)))
+			err = errors.Errorf("stats loading panicked: %v", r)
 		}
 	}()
-	h.getFreshStatsReader(readerCtx, ctx)
-	task, err := h.drainColTask(exit)
-	if err != nil {
-		if err != errExit {
-			logutil.BgLogger().Error("Fail to drain task for stats loading.", zap.Error(err))
+	if lastTask == nil {
+		task, err = h.drainColTask(exit)
+		if err != nil {
+			if err != errExit {
+				logutil.BgLogger().Error("Fail to drain task for stats loading.", zap.Error(err))
+			}
+			return task, err
 		}
```

Comment:
The `drainColTask` would not meet the case that both error and task is not nil?

Reply:
The implementation of drainColTask promises that it will not return a task together with error. Let me add more comments.

Reply:
Done